File: src/day09/main.py
from itertools import *


def sign(x): return 1 if x > 0 else (0 if x == 0 else -1)


def follow(head, tail):
    if all(head[axis]-1 <= tail[axis] <= head[axis]+1 for axis in range(2)):
        return
    for axis in range(2):
        tail[axis] += sign(head[axis]-tail[axis])


def propagate(nodes):
    for i in range(len(nodes)-1):
        follow(*nodes[i:i+2])


def sim_rope(nodes, direction, steps, visited1, visited9):
    dx, dy = 0, 0
    match direction:
        case "U": dy = 1
        case "D": dy = -1
        case "L": dx = -1
        case "R": dx = 1
    for _ in range(steps):
        nodes[0][0] += dx
        nodes[0][1] += dy
        propagate(nodes)
        visited1.add(tuple(nodes[1]))
        visited9.add(tuple(nodes[9]))


def main(input):
    state = [[0, 0] for _ in range(10)]
    visited1 = set([tuple(state[1])])
    visited9 = set([tuple(state[9])])
    for l in input.splitlines():
        direction, distance = l.split()
        sim_rope(state, direction, int(distance), visited1, visited9)
    print(f"part 1: {len(visited1)}")
    print(f"part 2: {len(visited9)}")

File: src/day09/main.rs
#![allow(unused_imports)]
use std::{
    collections::HashSet,
    iter::{self, FromIterator},
};

use itertools::Itertools;
use shared::puzzle_input;

fn follow(head: &[isize; 2], tail: &mut [isize; 2]) {
    if !(0..2)
        .map(|axis| ((head[axis] - 1)..=(head[axis] + 1)).contains(&tail[axis]))
        .all(|x| x)
    {
        for axis in 0..2 {
            tail[axis] += (head[axis] - tail[axis]).signum();
        }
    }
}

fn propagate(nodes: &mut [[isize; 2]]) {
    for i in 0..nodes.len() - 1 {
        follow(&nodes[i].clone(), &mut nodes[i + 1]);
    }
}

pub fn main() {
    let input = puzzle_input!();
    let mut nodes = [[0isize, 0isize]; 10];
    let mut visits1: HashSet<_> = HashSet::from_iter(iter::once([0isize; 2]));
    let mut visits9: HashSet<_> = HashSet::from_iter(iter::once([0isize; 2]));
    for line in input.split('\n') {
        let (dir, dist) = line.split_at(2);
        let dist = dist.parse().unwrap();
        let (dx, dy) = match dir {
            "U " => (0, 1),
            "D " => (0, -1),
            "L " => (-1, 0),
            "R " => (1, 0),
            _ => panic!(),
        };
        for _ in 0..dist {
            nodes[0][0] += dx;
            nodes[0][1] += dy;
            propagate(&mut nodes);
            visits1.insert(nodes[1]);
            visits9.insert(nodes[9]);
        }
    }

    println!("part 1: {}\npart 2: {}", visits1.len(), visits9.len());
}
